### ISSUE TITLE
Get nan while training

### ISSUE THREAD

**Issue by were** [dmlc/mxnet#942]
I wrote a mxnet-code to build MLP, but I get `nan` while training...
My code works well on other trainings, but failed on this.

`````
import mxnet as mx
import numpy
import sys
import logging
import math
import random

logging.basicConfig(level = logging.DEBUG)

class MyMetric(mx.metric.EvalMetric):
    def __init__(self):
        self.name = 'MSE'
        self.reset()

    def reset(self):
        self.res = 0.

    def eval_it(self, label, pred):
        pred = pred.asnumpy()
        label = label.asnumpy()
        res = 0
        #cnt = 0
        for i, j in zip(label, pred):
            delta = ((j - i) * (j - i)).sum() / i.shape[0]
            res += math.sqrt(delta)
            #cnt += 1 print cnt, i, j
        return res

    def update(self, label, pred):
        assert len(label) == len(pred)
        for i, j in zip(label, pred):
            self.res += self.eval_it(i, j)

    def get(self):
        return (self.name, self.res)

args = sys.argv[1:]
try:
    print args
    n_layers = int(args[0])
except:
    print 'The first argument should be number of layers!'
    quit()

nn = None
try:
    nn = mx.symbol.Variable('data')
    print 'shape is', nn.infer_shape
except:
    print 'Failed to initialize input layer!'
    quit()

try:
    for i in range(2, n_layers):
        nn = mx.symbol.FullyConnected(data = nn, name = 'hidden-recept-' + str(i), num_hidden = int(args[i]))
        nn = mx.symbol.Activation(data = nn, name = 'hidden-act-' + str(i), act_type = 'softrelu')
        #if random.randint(0, 1):
            #nn = mx.symbol.Dropout(data = nn, name = 'hidden-dropout', p = random.random())
except:
    print 'Failed to initailize hidden layer!'
    quit()

try:
    nn = mx.symbol.FullyConnected(data = nn, name = 'hidden-recept-' + str(n_layers), num_hidden = int(args[n_layers]),
            no_bias = True)
    nn = mx.symbol.LinearRegressionOutput(data = nn, name = 'softmax')
except:
    print 'Failed to build output layer!'
    quit()

BATCH_SIZE = 2048

data = open('data').readlines()
x,y = [],[]
for i in xrange(len(data)):
    if i % 2 == 0:
        x.append(map(float, data[i].split()))
    else:
        y.append(map(float, data[i].split()))
x = mx.ndarray.array(x).asnumpy()
y = mx.ndarray.array(y).asnumpy()
data_iter = mx.io.NDArrayIter(data = x, label = y, batch_size = BATCH_SIZE)

#print iterator.provide_data, iterator.provide_label

data = open('test').readlines()
x,y = [],[]
magnitude = 0.
for i in xrange(len(data)):
    tmp = map(float, data[i].split())
    if i % 2 == 0:
        x.append(tmp)
    else:
        y.append(tmp)
x = mx.ndarray.array(x).asnumpy()
y = mx.ndarray.array(y).asnumpy()
test_iter = mx.io.NDArrayIter(data = x, label = y, batch_size = BATCH_SIZE)
#print provide_data, provide_label

mon = mx.monitor.Monitor(1)
#mon = None

#mx.visualization.plot_network(nn)

model = mx.model.FeedForward(
        symbol = nn,
        ctx = mx.gpu(0),
        num_epoch = 1000,
        optimizer = mx.optimizer.RMSProp(learning_rate = 1e-2),
        initializer = mx.initializer.Xavier(magnitude = .3)
)

model.fit(
        X = data_iter,
        monitor = mon,
        eval_data = test_iter,
        eval_metric = MyMetric(),
        batch_end_callback = mx.callback.Speedometer(BATCH_SIZE, BATCH_SIZE)
)

model.save("res")

test_iter.reset()
for i in model.predict(test_iter):
    print i
`````

use `python builder.py 5 2 4 4 4 2` to build a 5-layer MLP, which 4 neurons in each hidden layer.
`data` and `test` are files of training and testing data.
both `test_iter` and `data_iter` are iterator for two `2*n NDArray`.

I tried to add a monitor to see what happened.
It told me the training process runs well on first 4 batches, and starts to be `nan` since 5.

``````
INFO:root:Batch:       1 hidden-recept-2_output         0.165509
INFO:root:Batch:       1 hidden-act-2_output            0.773078
INFO:root:Batch:       1 hidden-recept-3_output         0.145818
INFO:root:Batch:       1 hidden-act-3_output            0.722256
INFO:root:Batch:       1 hidden-recept-4_output         0.280935
INFO:root:Batch:       1 hidden-act-4_output            0.832174
INFO:root:Batch:       1 hidden-recept-5_output         0.427171
INFO:root:Batch:       1 softmax_output                 0.427171
INFO:root:Batch:       1 softmax_backward_data          0.518878
INFO:root:Batch:       1 softmax_backward_label         inf
INFO:root:Batch:       1 hidden-recept-5_backward_data  0.132449
INFO:root:Batch:       1 hidden-recept-5_backward_weight 795.523
INFO:root:Batch:       1 hidden-act-4_backward_data     0.0738911
INFO:root:Batch:       1 hidden-recept-4_backward_data  0.0274807
INFO:root:Batch:       1 hidden-recept-4_backward_weight 96.4226
INFO:root:Batch:       1 hidden-recept-4_backward_bias  133.524
INFO:root:Batch:       1 hidden-act-3_backward_data     0.0146831
INFO:root:Batch:       1 hidden-recept-3_backward_data  0.00441014
INFO:root:Batch:       1 hidden-recept-3_backward_weight 20.4973
INFO:root:Batch:       1 hidden-recept-3_backward_bias  26.6551
INFO:root:Batch:       1 hidden-act-2_backward_data     0.00228981
INFO:root:Batch:       1 hidden-recept-2_backward_data  0.518878
INFO:root:Batch:       1 hidden-recept-2_backward_weight 2.26191
INFO:root:Batch:       1 hidden-recept-2_backward_bias  4.22354
INFO:root:Batch:       1 data                           0.637145
INFO:root:Batch:       1 hidden-recept-2_weight         0.139162
INFO:root:Batch:       1 hidden-recept-2_bias           0.0458743
INFO:root:Batch:       1 hidden-recept-3_weight         0.125683
INFO:root:Batch:       1 hidden-recept-3_bias           0.0458821
INFO:root:Batch:       1 hidden-recept-4_weight         0.164311
INFO:root:Batch:       1 hidden-recept-4_bias           0.0458831
INFO:root:Batch:       1 hidden-recept-5_weight         0.191803
INFO:root:Batch:       1 softmax_label                  0.911673
INFO:root:Batch:       2 hidden-recept-2_output         0.158495
INFO:root:Batch:       2 hidden-act-2_output            0.75144
INFO:root:Batch:       2 hidden-recept-3_output         0.120121
INFO:root:Batch:       2 hidden-act-3_output            0.6759
INFO:root:Batch:       2 hidden-recept-4_output         0.200911
INFO:root:Batch:       2 hidden-act-4_output            0.764085
INFO:root:Batch:       2 hidden-recept-5_output         0.403537
INFO:root:Batch:       2 softmax_output                 0.403537
INFO:root:Batch:       2 softmax_backward_data          0.438727
INFO:root:Batch:       2 softmax_backward_label         0.0743718
INFO:root:Batch:       2 hidden-recept-5_backward_data  0.103892
INFO:root:Batch:       2 hidden-recept-5_backward_weight 587.519
INFO:root:Batch:       2 hidden-act-4_backward_data     0.0559733
INFO:root:Batch:       2 hidden-recept-4_backward_data  0.0141433
INFO:root:Batch:       2 hidden-recept-4_backward_weight 61.7605
INFO:root:Batch:       2 hidden-recept-4_backward_bias  91.3926
INFO:root:Batch:       2 hidden-act-3_backward_data     0.00706452
INFO:root:Batch:       2 hidden-recept-3_backward_data  0.00205192
INFO:root:Batch:       2 hidden-recept-3_backward_weight 7.46835
INFO:root:Batch:       2 hidden-recept-3_backward_bias  9.99114
INFO:root:Batch:       2 hidden-act-2_backward_data     0.00110245
INFO:root:Batch:       2 hidden-recept-2_backward_data  0.438727
INFO:root:Batch:       2 hidden-recept-2_backward_weight 0.979812
INFO:root:Batch:       2 hidden-recept-2_backward_bias  1.78149
INFO:root:Batch:       2 data                           0.639046
INFO:root:Batch:       2 hidden-recept-2_weight         0.156327
INFO:root:Batch:       2 hidden-recept-2_bias           0.11099
INFO:root:Batch:       2 hidden-recept-3_weight         0.135165
INFO:root:Batch:       2 hidden-recept-3_bias           0.0927643
INFO:root:Batch:       2 hidden-recept-4_weight         0.17869
INFO:root:Batch:       2 hidden-recept-4_bias           0.11218
INFO:root:Batch:       2 hidden-recept-5_weight         0.215187
INFO:root:Batch:       2 softmax_label                  0.909899
INFO:root:Batch:       3 hidden-recept-2_output         0.236518
INFO:root:Batch:       3 hidden-act-2_output            0.737678
INFO:root:Batch:       3 hidden-recept-3_output         0.28877
INFO:root:Batch:       3 hidden-act-3_output            0.693304
INFO:root:Batch:       3 hidden-recept-4_output         0.375687
INFO:root:Batch:       3 hidden-act-4_output            0.726632
INFO:root:Batch:       3 hidden-recept-5_output         0.494087
INFO:root:Batch:       3 softmax_output                 0.494087
INFO:root:Batch:       3 softmax_backward_data          0.349887
INFO:root:Batch:       3 softmax_backward_label         0.0550792
INFO:root:Batch:       3 hidden-recept-5_backward_data  0.0891329
INFO:root:Batch:       3 hidden-recept-5_backward_weight 396.782
INFO:root:Batch:       3 hidden-act-4_backward_data     0.0499654
INFO:root:Batch:       3 hidden-recept-4_backward_data  0.0180972
INFO:root:Batch:       3 hidden-recept-4_backward_weight 46.9512
INFO:root:Batch:       3 hidden-recept-4_backward_bias  67.844
INFO:root:Batch:       3 hidden-act-3_backward_data     0.00866769
INFO:root:Batch:       3 hidden-recept-3_backward_data  0.00255783
INFO:root:Batch:       3 hidden-recept-3_backward_weight 9.1302
INFO:root:Batch:       3 hidden-recept-3_backward_bias  12.3923
INFO:root:Batch:       3 hidden-act-2_backward_data     0.00135789
INFO:root:Batch:       3 hidden-recept-2_backward_data  0.349887
INFO:root:Batch:       3 hidden-recept-2_backward_weight 1.08412
INFO:root:Batch:       3 hidden-recept-2_backward_bias  1.87598
INFO:root:Batch:       3 data                           0.637617
INFO:root:Batch:       3 hidden-recept-2_weight         0.20213
INFO:root:Batch:       3 hidden-recept-2_bias           0.188711
INFO:root:Batch:       3 hidden-recept-3_weight         0.160229
INFO:root:Batch:       3 hidden-recept-3_bias           0.139786
INFO:root:Batch:       3 hidden-recept-4_weight         0.2199
INFO:root:Batch:       3 hidden-recept-4_bias           0.185218
INFO:root:Batch:       3 hidden-recept-5_weight         0.257822
INFO:root:Batch:       3 softmax_label                  0.90779
INFO:root:Batch:       4 hidden-recept-2_output         0.373694
INFO:root:Batch:       4 hidden-act-2_output            0.74138
INFO:root:Batch:       4 hidden-recept-3_output         0.495543
INFO:root:Batch:       4 hidden-act-3_output            0.790558
INFO:root:Batch:       4 hidden-recept-4_output         0.705683
INFO:root:Batch:       4 hidden-act-4_output            0.770608
INFO:root:Batch:       4 hidden-recept-5_output         0.688595
INFO:root:Batch:       4 softmax_output                 0.688595
INFO:root:Batch:       4 softmax_backward_data          0.270061
INFO:root:Batch:       4 softmax_backward_label         0.0496056
INFO:root:Batch:       4 hidden-recept-5_backward_data  0.0886355
INFO:root:Batch:       4 hidden-recept-5_backward_weight 232.365
INFO:root:Batch:       4 hidden-act-4_backward_data     0.0506027
INFO:root:Batch:       4 hidden-recept-4_backward_data  0.0245503
INFO:root:Batch:       4 hidden-recept-4_backward_weight 25.6307
INFO:root:Batch:       4 hidden-recept-4_backward_bias  32.7064
INFO:root:Batch:       4 hidden-act-3_backward_data     0.0113317
INFO:root:Batch:       4 hidden-recept-3_backward_data  0.00345556
INFO:root:Batch:       4 hidden-recept-3_backward_weight 7.18837
INFO:root:Batch:       4 hidden-recept-3_backward_bias  9.85327
INFO:root:Batch:       4 hidden-act-2_backward_data     0.00181982
INFO:root:Batch:       4 hidden-recept-2_backward_data  0.270061
INFO:root:Batch:       4 hidden-recept-2_backward_weight 1.21468
INFO:root:Batch:       4 hidden-recept-2_backward_bias  1.97961
INFO:root:Batch:       4 data                           0.641723
INFO:root:Batch:       4 hidden-recept-2_weight         0.267277
INFO:root:Batch:       4 hidden-recept-2_bias           0.272548
INFO:root:Batch:       4 hidden-recept-3_weight         0.201508
INFO:root:Batch:       4 hidden-recept-3_bias           0.194353
INFO:root:Batch:       4 hidden-recept-4_weight         0.274731
INFO:root:Batch:       4 hidden-recept-4_bias           0.258076
INFO:root:Batch:       4 hidden-recept-5_weight         0.303316
INFO:root:Batch:       4 softmax_label                  0.892841
INFO:root:Batch:       5 hidden-recept-2_output         0.534661
INFO:root:Batch:       5 hidden-act-2_output            0.777116
INFO:root:Batch:       5 hidden-recept-3_output         0.749784
INFO:root:Batch:       5 hidden-act-3_output            0.957261
INFO:root:Batch:       5 hidden-recept-4_output         1.13733
INFO:root:Batch:       5 hidden-act-4_output            0.920114
INFO:root:Batch:       5 hidden-recept-5_output         0.946593
INFO:root:Batch:       5 softmax_output                 0.946593
INFO:root:Batch:       5 softmax_backward_data          nan
INFO:root:Batch:       5 softmax_backward_label         0.0500394
INFO:root:Batch:       5 hidden-recept-5_backward_data  nan
INFO:root:Batch:       5 hidden-recept-5_backward_weight nan
INFO:root:Batch:       5 hidden-act-4_backward_data     nan
INFO:root:Batch:       5 hidden-recept-4_backward_data  nan
INFO:root:Batch:       5 hidden-recept-4_backward_weight nan
INFO:root:Batch:       5 hidden-recept-4_backward_bias  nan
INFO:root:Batch:       5 hidden-act-3_backward_data     nan
INFO:root:Batch:       5 hidden-recept-3_backward_data  nan
INFO:root:Batch:       5 hidden-recept-3_backward_weight nan
INFO:root:Batch:       5 hidden-recept-3_backward_bias  nan
INFO:root:Batch:       5 hidden-act-2_backward_data     nan
INFO:root:Batch:       5 hidden-recept-2_backward_data  nan
INFO:root:Batch:       5 hidden-recept-2_backward_weight nan
INFO:root:Batch:       5 hidden-recept-2_backward_bias  nan
INFO:root:Batch:       5 data                           0.640956
INFO:root:Batch:       5 hidden-recept-2_weight         nan
INFO:root:Batch:       5 hidden-recept-2_bias           nan
INFO:root:Batch:       5 hidden-recept-3_weight         nan
INFO:root:Batch:       5 hidden-recept-3_bias           nan
INFO:root:Batch:       5 hidden-recept-4_weight         nan
INFO:root:Batch:       5 hidden-recept-4_bias           nan
INFO:root:Batch:       5 hidden-recept-5_weight         nan
INFO:root:Batch:       5 softmax_label                  nan
INFO:root:Batch:       6 hidden-recept-2_output         nan
``````

**Comment by mli** [comment]
try `clip_gradient? 

https://github.com/dmlc/mxnet/blob/master/python/mxnet/optimizer.py#L435

**Comment by were** [comment]
How much will be proper?

**Comment by mli** [comment]
i suggest to start with 10

**Comment by were** [comment]
It doesn't help.
Now it starts to be `nan` in batch one.

**Comment by mli** [comment]
have you tried different parameters, such as smaller learning rate, or optimizer.SGD

**Comment by were** [comment]
Yep, I tried 1e-4 learning rate. Also it doesn't help, too.

**Comment by were** [comment]
I think the key problem is that the code starts to give `nan` after the first batch was evaluated by the metric.

Besides, I think `RMSProp` is more suitable for my work, via my test.

**Comment by were** [comment]
Sorry, it's my fault. There is something wrong with my data.

**Comment by were** [comment]
Close the issue.
